sdks/java/io/rrio/src/test/java/org/apache/beam/io/requestresponse/ThrottleTest.java

```diff
@@ -0,0 +1,354 @@
+              assertTimestampIntervalsMatch(timestamps, lessThan(expectedInterval));
+              return null;
+            });
+
+    pipeline.run();
+  }
+
+  /** Tests whether a pulse of elements totaled greater than the maximum rate are throttled. */
+  @Test
+  public void givenElementSizeExceedsRate_thenEmitsAtRate() {
+    Rate rate = Rate.of(1, Duration.standardSeconds(1L));
+    long expectedMillis = rate.getInterval().getMillis();
+    // tolerate 5% error.
+    long toleratedError = (long) (0.05 * (double) expectedMillis);
+    Duration expectedInterval = Duration.millis(expectedMillis - toleratedError);
+    List<Integer> items = Stream.iterate(0, i -> i + 1).limit(3).collect(Collectors.toList());
+    PCollection<Integer> throttled = pipeline.apply(Create.of(items)).apply(transformOf(rate));
+
+    PAssert.that(throttled).containsInAnyOrder(items);
+    PAssert.that(timestampsOf(throttled))
+        .satisfies(
+            itr -> {
+              List<Instant> timestamps =
+                  StreamSupport.stream(itr.spliterator(), true)
+                      .sorted()
+                      .collect(Collectors.toList());
+              assertTimestampIntervalsMatch(timestamps, greaterThan(expectedInterval));
+              return null;
+            });
+
+    pipeline.run();
+  }
+
+  @Test
+  public void givenLargerElementSize_noDataLost() {
+    Rate rate = Rate.of(1_000, Duration.standardSeconds(1L));
+    List<Integer> items = Stream.iterate(0, i -> i + 1).limit(3_000).collect(Collectors.toList());
+    PCollection<Integer> throttled = pipeline.apply(Create.of(items)).apply(transformOf(rate));
+
+    PAssert.that(throttled).containsInAnyOrder(items);
+
+    pipeline.run();
+  }
+
+  /** Tests withMetricsCollected that Counters populate appropriately. */
+  @Test
+  public void givenCollectMetricsTrue_thenPopulatesMetrics() {
+    long size = 300;
+    Rate rate = Rate.of(100, Duration.standardSeconds(1L));
+
+    List<Integer> list = Stream.iterate(0, i -> i + 1).limit(size).collect(Collectors.toList());
+
+    pipeline.apply(Create.of(list)).apply(transformOf(rate).withMetricsCollected());
+
+    PipelineResult pipelineResult = pipeline.run();
+    pipelineResult.waitUntilFinish();
+    MetricResults results = pipelineResult.metrics();
+
+    assertThat(getCount(results, INPUT_ELEMENTS_COUNTER_NAME), equalTo(size));
+    assertThat(getCount(results, OUTPUT_ELEMENTS_COUNTER_NAME), equalTo(size));
+  }
+
+  /** Tests that an upstream GlobalWindows gets reapplied to the resulting PCollection. */
+  @Test
+  public void givenUpstreamGlobalWindows_thenReassignedToGlobalWindow() {
+    PCollection<Integer> throttled =
+        pipeline
+            .apply(Create.of(1, 2, 3))
+            .apply(Window.into(new GlobalWindows()))
+            .apply(transformOf(Rate.of(1, Duration.standardSeconds(1L))));
+
+    assertThat(throttled.getWindowingStrategy().getWindowFn(), equalTo(new GlobalWindows()));
+
+    pipeline.run();
+  }
```

Comment:
I don't think these tests do quite enough. They show that the windowing strategy is reapplied, but they don't show that elements maintain their original window (which is probably the more important piece). I think a better test would be something like:

```
pipeline.apply(<create elements 1-10>)
  .apply(<map fn that gives timestamps equal to starttime + (element * 1 second)>)
  .apply(<fixed window of length 5 seconds>)
  .apply(<throttle with rate of 1 element every 5 seconds>)
  .apply(<combine - count per window>)
 
<assert that there are 2 elements output with counts of 5>
```

One of my core concerns with the previous approach (haven't fully reviewed the new one) is that we're losing the info we need to do accurate windowing.

Reply:
Good point. I was actually considering inspecting the BoundedWindow and PaneInfo available in the DoFn but I like your idea as well. I'll figure something out. Thank you for this.

Reply:
@damondouglas it looks like this use case still isn't tested, would you mind covering it? If it is and I just missed it, please point me to the test case

Reply:
@damccorm I added `givenUpstreamFixedWindow_thenPreservesWindowAssignments` and `givenGlobalUpstreamWindow_thenCanApplyDownstreamWindow` to `ThrottleTest`.

Reply:
I don't think either of those test what I'm talking about here. They test that the correct windows are applied, but they don't test that the elements maintain their original (correct) window assignments. For example:

```
  @Test
  public void givenGlobalUpstreamWindow_thenCanApplyDownstreamWindow() {
    Rate rate = Rate.of(3, Duration.standardSeconds(1L));

    PCollection<Integer> unthrottled =
        pipeline.apply(
            Create.timestamped(
                TimestampedValue.of(0, epochPlus(0L)),
                TimestampedValue.of(1, epochPlus(1_000L)),
                TimestampedValue.of(2, epochPlus(2_000L))));

    PAssert.that(unthrottled).inWindow(GlobalWindow.INSTANCE);

    PCollection<Integer> throttled =
        unthrottled
            .apply(transformOf(rate))
            .apply(Window.into(FixedWindows.of(Duration.standardSeconds(5L))));
    PAssert.that(throttled).inWindow(new IntervalWindow(Instant.EPOCH, epochPlus(5_000L)));

    pipeline.run();
  }
```

asserts `.inWindow(new IntervalWindow(Instant.EPOCH, epochPlus(5_000L)));`, but it doesn't actually assert that all 3 values are in their correct interval windows (both tests are useful/important, but the latter is what I'm worried about in this PR)